### PR TITLE
Enable C3, H2, S2 and S3 HIL

### PR DESCRIPTION
**Pull request by SergioGasquez**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
- Enables C3, H2, S2 and S3 HIL.
- Updates the version of `probe-rs` being used.
##### Limitations
 Some test are disabled for some targets, see: https://github.com/esp-rs/esp-hal/issues/1524 and https://github.com/esp-rs/esp-hal/issues/1509

> [!IMPORTANT]  
> Once merged we need to add the `HIL Test | esp32c3`, `HIL Test | esp32h2`,  `HIL Test | esp32s2` and `HIL Test | esp32s3` test to the required checks

#### Testing
Manually triggered the HIL tests workflow: https://github.com/esp-rs/esp-hal/actions/runs/8891148274

### CODE REVIEW

File: hil-test/tests/spi_full_duplex_dma.rs [hunk 1]
```diff
@@ -121,6 +121,10 @@ mod tests {
         let transfer = spi.dma_transfer(&mut send, &mut receive).unwrap();
         transfer.wait().unwrap();
         assert_eq!(send[0], receive[0]);
+        assert_eq!(send[1], receive[1]);
+        // Read the 2 remaining bytes so it does not cause issues in the next test.
```
Comment (bjoernQ) [inline hil-test/tests/spi_full_duplex_dma.rs:125]: what issues does it cause if we don't repeat the transfer?

Comment (bjoernQ) [reply hil-test/tests/spi_full_duplex_dma.rs:125]: that's very weird 🤷‍♂️

**Review by bjoernQ — APPROVED** [review]
LGTM - however I'm a noob in regards to CI

File: hil-test/tests/spi_full_duplex_dma.rs (hunk 1 above)
Comment (jessebraham) [reply hil-test/tests/spi_full_duplex_dma.rs:125]: This test was passing before, it should not need to be changed. This is either a bug for the S2 implementation or an issue with the new runners, I'd imagine. Not really excited about merging hacks into our HIL tests...

Comment (SergioGasquez) [reply hil-test/tests/spi_full_duplex_dma.rs]: This should not be related to the new runner, the old runner also required this change and the same behavior happens locally. I've reverted the changes and if we merge this, we should open an issue to investigate `spi_fullduplex_dma` test on S2 which fails in 2 different ways, and its currently disabled.

Comment (jessebraham) [reply hil-test/tests/spi_full_duplex_dma.rs]: I would prefer all HIL tests to be passing for all devices before merging this personally, but sure whatever works.

**Comment by SergioGasquez** [comment]
Converted to draft as `uart` is now failing for C3 after the rebase, will investigate the issue

**Comment by SergioGasquez** [comment]
Disabled the following test for the following targets:
- `delay`: H2 and S2
- `spi_full_duplex_dma`: S2

All the issues are covered in https://github.com/esp-rs/esp-hal/issues/1509 and https://github.com/esp-rs/esp-hal/issues/1524

HIL are now running green: https://github.com/esp-rs/esp-hal/actions/runs/8891148274

File: hil-test/tests/spi_full_duplex_dma.rs [hunk 2]
```diff
@@ -8,7 +8,8 @@
 //!
 //! Connect MISO (GPIO2) and MOSI (GPIO4) pins.
 
-//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s2 esp32s3
+//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s3
+// TODO: esp32s2 is disabled due to https://github.com/esp-rs/esp-hal/issues/1524
```
Comment (jessebraham) [inline hil-test/tests/spi_full_duplex_dma.rs:12]: Same as above.

File: hil-test/tests/delay.rs [hunk 3]
```diff
@@ -1,5 +1,9 @@
 //! Delay Test
 
+//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32s3
+// TODO: esp32h2 is disabled due to https://github.com/esp-rs/esp-hal/issues/1509
```
Comment (jessebraham) [inline hil-test/tests/delay.rs:4]: I don't think we need a TODO comment and an issue, we've been working on eliminating TODO comments.

File: .github/workflows/hil.yml [hunk 4]
```diff
@@ -95,24 +98,45 @@ jobs:
...
     steps:
       - uses: actions/download-artifact@v4
         with:
           name: tests-${{ matrix.target.soc }}
-          path: tests
-      - name: Run tests
+          path: tests-${{ matrix.target.soc }}
+
+      - name: Run Tests
         run: |
```
Comment (jessebraham) [inline .github/workflows/hil.yml:119]: Would really like to see this implemented in the `xtask` instead, please.

Comment (SergioGasquez) [reply .github/workflows/hil.yml:119]: Me too, this bash code and the one preparing the artifacts looks quite bad to me. Any suggestions on how to integrate this with `xtask` other than creating a subcommand that points to a folder, and it executes all the elfs on that folder?

**Review by JurajSadel — COMMENTED** [review]
LGTM, left one nitpick comment, sorry. Not going to approve so it's not accidentally merged before remaining discussion threads are resolved.

File: hil-test/README.md [hunk 5]
```diff
@@ -61,19 +59,24 @@ Our Virtual Machines have the following setup:
...
   - Devkit: `ESP32-C3-DevKit-RUST-1` connected via USB-Serial-JTAG.
     - `GPIO2` and `GPIO4` are connected.
-  - VM: Ubuntu 20.04.5 configured with the following [setup](#vm-setup)
+  - RPi: Raspbian 12 configured with the following [setup](#vm-setup)
 - ESP32-C6 (`esp32c6-usb`):
   - Devkit: `ESP32-C6-DevKitC-1 V1.2` connected via USB-Serial-JTAG (`USB` port).
     - `GPIO2` and `GPIO4` are connected.
   - RPi: Raspbian 12 configured with the following [setup](#vm-setup)
-- ESP32-H2 (`esp32h2-usb`):
+- ESP32-H2 (`esp32h2-usb`): Currently disabled
```
Comment (JurajSadel) [inline hil-test/README.md:67]: Nitpick: This comment might be misleading, some tests are allowed on H2.

Comment (SergioGasquez) [reply hil-test/README.md]: Ups, good catch! I forgot to remove this as H2 was added later on

File: .github/workflows/hil.yml (hunk 4 above)
Comment (SergioGasquez) [reply .github/workflows/hil.yml:120]: Added a subcommand to run all the elfs in a folder in [`daf8746`](https://github.com/esp-rs/esp-hal/pull/1513/commits/daf874691664ef5ecfdb3f2c9f09579120213534). Here is a HIL run: https://github.com/esp-rs/esp-hal/actions/runs/8924001488 with that changes, it increases the time it takes for the run as it needs to build the xtask package.

**Review by jessebraham — APPROVED** [review]
LGTM, thanks!

**Pull request merged**